### PR TITLE
Compiler flag adjustments for Intel Compiler

### PR DESCRIPTION
**Pull request by umar456**
The Intel compiler uses fast math by default in order to perform some optimization. These
optimizations causes some of our tests to fail. This PR handles several of the issues 
brought up with the new compiler

Description
-----------
* Add AF_WITH_FAST_MATH to options which enable the use of fast math flags in all compilers. This is off by default
* Fix several warnings with clang and icx
* Refactor GitHub actions so that the builds only execute if clang-format and docs pass
* Add additional compilers to GitHub actions

Changes to Users
----------------
* Users can now enable AF_WITH_FAST_MATH for additional performance at the cost of precision

Checklist
---------

- [x] Rebased on latest master
- [x] Code compiles
- [x] Tests pass
- ~[ ] Functions added to unified API~
- ~[ ] Functions documented~

### CODE REVIEW

File: src/backend/common/TemplateArg.hpp [hunk 1]
```diff
@@ -16,7 +16,7 @@
 #include <utility>
 
 template<typename T>
-class TemplateTypename;
+struct TemplateTypename;
```
Comment (9prady9) [inline src/backend/common/TemplateArg.hpp:19]: good catch :)

File: test/testHelpers.hpp [hunk 2]
```diff
@@ -223,9 +223,17 @@ bool noDoubleTests(af::dtype ty);
...
+#define SUPPORTED_TYPE_CHECK(type)                                \
+    if (noDoubleTests((af_dtype)af::dtype_traits<type>::af_type)) \
+        GTEST_SKIP() << "Device doesn't support Doubles";         \
+    if (noHalfTests((af_dtype)af::dtype_traits<type>::af_type))   \
+        GTEST_SKIP() << "Device doesn't support Half";
+
+#ifdef AF_WITH_FAST_MATH
+#define NOFASTMATH() GTEST_SKIP() << "ArrayFire compiled with AF_WITH_FAST_MATH"
+#else
+#define NOFASTMATH()
```
Comment (9prady9) [inline test/testHelpers.hpp:235]: I would say `SKIP_TEST_IF_FAST_MATH_ENABLED` is more clear on what the macro accomplishes. It is a bit long but clear I think.

**Review by 9prady9 — COMMENTED** [review]
did a quick pass, lgtm

**Comment by umar456** [comment]
Addressed all feedback.

**Review by syurkevi — APPROVED** [review]

**Pull request merged**